### PR TITLE
Feat/custom-select

### PR DESCRIPTION
**Pull request by RostiMelk**

### CODE REVIEW

File: src/stories/Forms/Select/Select.tsx [hunk 1]
```diff
@@ -0,0 +1,489 @@
...
+  const [selected, setSelected] = useState<string | null>(value);
+  const [selectListHeight, setSelectListHeight] = useState(0);
+  const [filterSearch, setFilterSearch] = useState('');
+
+  useEffect(() => {
+    setSelected(value);
+  }, [value]);
+
+  useEffect(() => {
+    const handleKeyDown = (e: KeyboardEvent) => {
```
Comment (amfestoy) [inline src/stories/Forms/Select/Select.tsx:309]: Move this function definition out of the useEffect for better readability 👍

File: src/stories/Forms/Select/Select.tsx [hunk 2]
```diff
@@ -0,0 +1,489 @@
...
+
+    window.addEventListener('keydown', handleKeyDown);
+
+    return () => {
+      window.removeEventListener('keydown', handleKeyDown);
+    };
+  }, [isOpen]);
+
+  useEffect(() => {
+    setSelectListHeight(getSelectListHeight(showItems, selectListRef));
```
Comment (amfestoy) [inline src/stories/Forms/Select/Select.tsx:376]: Could maybe setSelecListHeight only be done on the initial render?

File: src/stories/Forms/Select/Select.tsx [hunk 3]
```diff
@@ -0,0 +1,489 @@
...
+    if (!selectedValue) {
+      console.error('Select component: value is not defined', target);
+      return;
+    }
+    setFilterSearch('');
+    setIsOpen(false);
+    setSelected(selectedValue);
+    selectButtonRef?.current?.focus();
+    onChange &&
+      onChange({ target: { name: selectName, id, value: selectedValue } } as React.ChangeEvent<HTMLInputElement>);
```
Comment (amfestoy) [inline src/stories/Forms/Select/Select.tsx:394]: Note to self, discuss onChange events with Rosti

File: src/stories/Forms/Select/Select.tsx [hunk 4]
```diff
@@ -0,0 +1,489 @@
...
+
+  const selectContent = getLabelOfSelectedValue(children, selected);
+  const selectContentAsString = getSelectContentAsString(selectContent);
+  const filteredChildren = getChildrenByQuery(children, filterSearch);
+  const filteredChildrenEmpty = React.Children.toArray(filteredChildren).length === 0;
+
+  return (
+    <Wrapper className={className} ref={selectWrapperRef} onBlur={handleBlur}>
+      <SelectButton
+        aria-activedescendant={selected}
```
Comment (amfestoy) [inline src/stories/Forms/Select/Select.tsx:410]: Your aria knowledge and code is beautiful as always 👏

File: src/stories/Forms/Select/Select.tsx [hunk 5]
```diff
@@ -0,0 +1,489 @@
...
+          aria-expanded={isOpen}
+          aria-labelledby={ariaDescribedBy}
+          aria-multiselectable='false'
+          height={selectListHeight}
+          id={`${id}-select-list`}
+          isOpen={isOpen}
+          role='listbox'
+          tabIndex={isOpen ? 0 : -1}
+        >
+          <SelectListHoverBackground ref={selectListHoverBackgroundRef} aria-hidden />
```
Comment (amfestoy) [inline src/stories/Forms/Select/Select.tsx:458]: Note to self, explain me the need of a specific component to do hover background

File: .prettierrc [hunk 6]
```diff
@@ -3,5 +3,6 @@
   "semi": true,
   "jsxSingleQuote": true,
   "singleQuote": true,
-  "tabWidth": 2
+  "tabWidth": 2,
+  "printWidth": 120
```
Comment (amfestoy) [inline .prettierrc:7]: Could we do 100? I think personally 120 is hard to read

**Comment by amfestoy** [comment]
Easy to read code, which always make me happy! Lets discuss some points when you come to work 😊

File: src/stories/Forms/PhoneInput/index.tsx [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+    const cleanedValue = e.target.value.replace(/\D/g, ''); // remove all non-numeric characters
+    const phoneNumber = getAsYouType(regionCode).reset(cleanedValue); // format the phone number
+    setPhoneNumber(phoneNumber);
+  };
+
+  useEffect(function handleInitialRegionCode(): void {
+    if (regionCode) return;
+    if (geoLocate) {
+      geoLocateRegionCode().then((r) => setRegionCode(r));
+    } else {
```
[the remark this answers is not in the document]
Comment (RostiMelk) [reply src/stories/Forms/PhoneInput/index.tsx]: Honestly think it's more readable this way 🤔

File: src/stories/Forms/PhoneInput/index.tsx [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+      const firstCountryCode = Number(countryCodes[0]);
+      setRegionCode(getRegionCodeForCountryCode(firstCountryCode));
+    }
+  }, []);
+
+  useEffect(
+    function handlePlaceholder(): void {
```
[the remark this answers is not in the document]
Comment (RostiMelk) [reply src/stories/Forms/PhoneInput/index.tsx]: Got it from somewhere on Twitter, no clue where 😅

File: src/stories/Forms/Select/Select.tsx (hunk 2 above)
Comment (RostiMelk) [reply src/stories/Forms/Select/Select.tsx]: It has to change depending on the filter result and if children change dynamically.

File: src/stories/Forms/Select/Select.tsx (hunk 5 above)
Comment (RostiMelk) [reply src/stories/Forms/Select/Select.tsx]: Added a comment to the component.

File: src/stories/Forms/Select/Select.tsx (hunk 1 above)
Comment (RostiMelk) [reply src/stories/Forms/Select/Select.tsx]: Too many dependencies, it will be a mess.

File: src/stories/Forms/Select/Select.tsx (hunk 3 above)
Comment (RostiMelk) [reply src/stories/Forms/Select/Select.tsx]: As discussed: Returning a fake event target so the value is compatible with Formik.

**Pull request merged**